pkg/lockfile/extractor.go

```diff
@@ -0,0 +1,64 @@
+package lockfile
+
+import (
+	"io"
+	"os"
+	"path/filepath"
+)
+
+// DepFile is an abstraction for a file that has been opened for extraction,
+// and that knows how to open other DepFiles relative to itself.
+type DepFile interface {
+	io.Reader
+
+	// Open opens an NestedDepFile based on the path of the
+	// current DepFile if the provided path is relative.
+	//
+	// If the path is an absolute path, then it is opened absolutely.
+	Open(string) (NestedDepFile, error)
+
+	Path() string
+}
+
+// NestedDepFile is an abstraction for a file that has been opened while extracting another file,
+// and would need to be closed.
+type NestedDepFile interface {
+	io.Closer
+	DepFile
+}
+
+type Extractor interface {
+	// ShouldExtract checks if the Extractor should be used for the given path.
+	ShouldExtract(path string) bool
+	Extract(f DepFile) ([]PackageDetails, error)
```

Comment:
Could we add
```
var ErrNotSupported = errors.New("this file does not support opening files")
```

per https://github.com/google/osv-scanner/issues/176#issuecomment-1637238126  ?

Reply:
@another-rex